indico/modules/events/registration/models/registrations.py

```diff
@@ -488,6 +488,11 @@ def update_state(self, approved=None, paid=None, rejected=None, _skip_moderation
                 self.state = RegistrationState.pending
             elif paid is False and payment_required:
                 self.state = RegistrationState.unpaid
+        elif self.state == RegistrationState.rejected:
+            if rejected is False and payment_required is False and moderation_required:
+                self.state = RegistrationState.pending
+            if rejected is False and payment_required:
+                self.state = RegistrationState.unpaid
```

Comment:
Payment happens only after approval, so I don't think rejected=>unpaid can happen. Resetting rejected should always go back to pending. The only edge case where you would not go back to pending is if moderation has been disabled in the meantime - in that case you'd go to complete.